src/groups/mqb/mqbblp/mqbblp_relayqueueengine.cpp

```diff
@@ -1682,17 +1676,472 @@ bool RelayQueueEngine::subscriptionId2upstreamSubQueueId(
         const Routers::SubscriptionIds::SharedItem itId =
             d_queueState_p->routingContext().d_groupIds.find(subscriptionId);
         if (!itId) {
+            BMQ_LOGTHROTTLE_ERROR()
+                << "#QUEUE_UNKNOWN_SUBSCRIPTION_ID "
+                << "Remote queue: " << d_queueState_p->uri()
+                << " (id: " << d_queueState_p->id()
+                << ") received a PUSH message for guid " << msgGUID
+                << ", with unknown Subscription Id " << subscriptionId;
+
             return false;  // RETURN
         }
         if (itId->value().d_priorityGroup == 0) {
             // The 'd_queueContext.d_groupIds' may contain new ids for which
             // configure response is not received yet.
+
+            BMQ_LOGTHROTTLE_ERROR()
+                << "#QUEUE_UNKNOWN_SUBSCRIPTION_ID "
+                << "Remote queue: " << d_queueState_p->uri()
+                << " (id: " << d_queueState_p->id()
+                << ") received a PUSH message for guid " << msgGUID
+                << ", with unconfigured Subscription Id " << subscriptionId;
+
             return false;
         }
         *subQueueId = itId->value().upstreamSubQueueId();
     }
     return true;
 }
 
+unsigned int
+RelayQueueEngine::push(mqbi::StorageMessageAttributes*           attributes,
+                       const bmqt::MessageGUID&                  msgGUID,
+                       const bsl::shared_ptr<bdlbb::Blob>&       appData,
+                       const bmqp::Protocol::SubQueueInfosArray& subscriptions,
+                       bool                                      isOutOfOrder)
+{
+    if (isOutOfOrder) {
+        BSLS_ASSERT_SAFE(subscriptions.size() == 1);
+
+        // No guarantee of uniqueness.  Cannot use PushStream.
+        unsigned int upstreamSubQueueId;
+
+        if (subscriptionId2upstreamSubQueueId(msgGUID,
+                                              &upstreamSubQueueId,
+                                              subscriptions.begin()->id())) {
+            App_State* app = findApp(upstreamSubQueueId);
+
+            if (app == 0) {
+                BMQ_LOGTHROTTLE_ERROR()
+                    << "#QUEUE_UNKNOWN_SUBSCRIPTION_ID "
+                    << "Remote queue: " << d_queueState_p->uri()
+                    << " (id: " << d_queueState_p->id()
+                    << ") discarding a PUSH message for guid " << msgGUID
+                    << ", with unknown App Id " << upstreamSubQueueId;
+
+                return 0;  // RETURN
```

Comment:
The return values of this function are a little opaque to me, and don't seem to be documented in the header file.

Reply:
Adding
```
/// Return number of inserted PushStream Elements.
```